android/app/src/main/java/co/edgesecure/app/MessagesWorker.java

```diff
@@ -0,0 +1,268 @@
+            .setContentTitle(title)
+            .setContentText(message)
+            .setContentIntent(pendingIntent)
+            .setSmallIcon(R.mipmap.edge_logo_hollow)
+            .build();
+    notificationManager.notify(notificationId, notification);
+  }
+
+  /** Hides our local notification. */
+  private void cancelNotification() {
+    Context context = getApplicationContext();
+    NotificationManager notificationManager =
+        (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
+
+    notificationManager.cancel(notificationId);
+  }
+
+  /** Schedules this worker to run periodically in the background, if necessary. */
+  public static void ensureScheduled(Context context) {
+    WorkManager manager = WorkManager.getInstance(/* context */ );
+
+    // Run periodically, but only with networking:
+    Constraints constraints =
+        new Constraints.Builder().setRequiredNetworkType(NetworkType.CONNECTED).build();
+    PeriodicWorkRequest messagesWorkRequest =
+        new PeriodicWorkRequest.Builder(MessagesWorker.class, 12, TimeUnit.HOURS)
+            .setConstraints(constraints)
+            .build();
+
+    // Add the task if it doesn't exist already:
+    manager.enqueueUniquePeriodicWork(workId, ExistingPeriodicWorkPolicy.KEEP, messagesWorkRequest);
+  }
+
+  /** Schedules this worker to run once right away, for debugging. */
+  public static void testRun(Context context) {
+    WorkManager manager = WorkManager.getInstance(/* context */ );
+
+    // Reset anything that might interfere with the test:
+    manager.cancelAllWork();
+    manager.pruneWork();
+
+    // Schedule ourselves to run once:
+    OneTimeWorkRequest messagesWorkRequest =
+        new OneTimeWorkRequest.Builder(MessagesWorker.class).build();
+    manager.enqueue(messagesWorkRequest);
+  }
+
+  /** Does a request / reply with the auth server. */
+  public String authFetch(String uri, String body) throws IOException {
+    HttpsURLConnection connection = null;
+    try {
+      // Set up the HTTPS connection:
+      connection = (HttpsURLConnection) new URL(uri).openConnection();
+      SSLContext context = SSLContext.getInstance("TLS");
+      context.init(null, null, null);
+      connection.setSSLSocketFactory(context.getSocketFactory());
+
+      // Add the auth server headers:
+      byte[] bodyData = body.getBytes("UTF-8");
+      connection.setRequestProperty("Accept", "application/json");
+      connection.setRequestProperty("Content-Type", "application/json");
+      connection.setRequestProperty("Authorization", "Token " + EdgeApiKey.apiKey);
+      connection.setRequestProperty("Content-Length", Integer.toString(bodyData.length));
+      connection.setRequestMethod("POST");
+      connection.setDoInput(true);
+      connection.setDoOutput(true);
+      connection.setUseCaches(false);
+
+      // Send the body:
+      OutputStream wr = connection.getOutputStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *wr*. The resource is referenced at lines 238, 239. The resource is closed at line 240. There are other execution paths that do not contain closure statements, for example, when OutputStream.write throws an exception. Ensure *wr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).